### DISCUSSION TITLE
可以给CRUDEnable增加 #[serde(flatten)]这个支持吗？

### DISCUSSION THREAD

**Discussion by peacess** [rbatis/rbatis#284] [General]
这样就相当于增加了嵌入结构体的支持了
#[derive(Serialize, Deserialize)]
pub struct One{
    pub name: String,
    pub code: String,
}
#[derive(Serialize, Deserialize, Clone, Debug, Default, CRUDEnable)]
pub struct Two {
    #[serde(flatten)]
    pub one: One,
    pub id: String,
}

**Comment by zhuxiujia** [comment]
> 这样就相当于增加了嵌入结构体的支持了
> #[derive(Serialize, Deserialize)]
> pub struct One{
> pub name: String,
> pub code: String,
> }
> #[derive(Serialize, Deserialize, Clone, Debug, Default, CRUDEnable)]
> pub struct Two {
> #[serde(flatten)]
> pub one: One,
> pub id: String,
> }

可以的 参见官网文档写法
,,,rust

//例子1(全部自动判断):
    #[crud_enable]
    #[derive(Serialize, Deserialize, Clone, Debug)]
    pub struct BizActivity {
        pub id: Option<String>,
        pub name: Option<String>,
        pub delete_flag: Option<i32>,
    }
// 例子2（只自定义表名，其他自动）:
    #[crud_enable(table_name:biz_activity)]
    #[derive(Serialize, Deserialize, Clone, Debug)]
    pub struct BizActivity {
        pub id: Option<String>,
        pub name: Option<String>,
        pub delete_flag: Option<i32>,
    }
// 例子3（全部自定义，其他自动）:
    #[crud_enable( id_name:id |  id_type:String |  table_name:biz_activity |  table_columns:id,name,delete_flag )]
    #[derive(Serialize, Deserialize, Clone, Debug)]
    pub struct BizActivity {
        pub id: Option<String>,
        pub name: Option<String>,
        pub delete_flag: Option<i32>,
    }

,,,

**Comment by pymongo** [comment]
> > 这样就相当于增加了嵌入结构体的支持了
> > #[derive(Serialize, Deserialize)]
> > pub struct One{
> > pub name: String,
> > pub code: String,
> > }
> > #[derive(Serialize, Deserialize, Clone, Debug, Default, CRUDEnable)]
> > pub struct Two {
> > #[serde(flatten)]
> > pub one: One,
> > pub id: String,
> > }
> 
> 可以的 参见官网文档写法 ,,,rust
> 
> //例子1(全部自动判断): #[crud_enable] #[derive(Serialize, Deserialize, Clone, Debug)] pub struct BizActivity { pub id: Option, pub name: Option, pub delete_flag: Option, } // 例子2（只自定义表名，其他自动）: #[crud_enable(table_name:biz_activity)] #[derive(Serialize, Deserialize, Clone, Debug)] pub struct BizActivity { pub id: Option, pub name: Option, pub delete_flag: Option, } // 例子3（全部自定义，其他自动）: #[crud_enable( id_name:id | id_type:String | table_name:biz_activity | table_columns:id,name,delete_flag )] #[derive(Serialize, Deserialize, Clone, Debug)] pub struct BizActivity { pub id: Option, pub name: Option, pub delete_flag: Option, }
> 
> ,,,

我试了下还是不行，原则上用宏的话 rbaits 的宏要在 serde 之后解析？否则下列写法生成的代码还是没法将 flatten 里面的字段读取出来

```rust
#[derive(rbatis::CRUDTable, Serialize, Deserialize, Debug)]
struct BaseFields {
    id: Id,
    created_at: i64,
    updated_at: i64,
    version: i32,
    is_deleted: bool
}

// #[rbatis_macro_driver::crud_table]
#[derive(rbatis::CRUDTable, Serialize, Deserialize, Debug)]
struct Orders {
    #[serde(flatten)]
    base: BaseFields,
    user_id: Id,
}
```

> error returned from database: column "base" does not exist

用 debug_mode 看宏生成的代码就发现，orders 表映射就没有 created_at 等字段

**Comment by zhuxiujia** [comment]
v4 支持任意serde的功能
